Create and parse vendor extension attribute

mariomaz opened the pull request:
See [PPM-136](https://jira.prplfoundation.org/browse/PPM-136)

mariomaz committed:
tlvf: add class for WSC Vendor Extension and subelements

Add new class for WSC Vendor Extension attribute (type 0x1049) using by
default Vendor ID 0x00372A (i.e.: WFA).

Add new structs for the WFA Vendor Extension subelements that we use:
Version2 (0x00) and Multi-AP Identifier (0x06).

Signed-off-by: Mario Maz <mmazzapater@gmail.com>

mariomaz committed:
agent: slave: remove unnecessary cast

Signed-off-by: Mario Maz <mmazzapater@gmail.com>

mariomaz committed:
tlvf: remove unused deprecated classes and structs

Remove the following unused deprecated classes and structs:
- sWscAttrVersion2
- cWscAttrVersion2
- sWscAttrVendorExtMultiAp
- cWscVendorExtWfa

Signed-off-by: Mario Maz <mmazzapater@gmail.com>

mariomaz committed:
tlvf: simplify cConfigData

Prepare cConfigData used in vendor specific message
ACTION_APMANAGER_WIFI_CREDENTIALS_UPDATE_REQUEST to replace the struct
`sWscAttrVendorExtMultiAp` with a single `bss_type` byte (this struct
is wrong and is going to be removed in a future commit).

Signed-off-by: Mario Maz <mmazzapater@gmail.com>

mariomaz committed:
tlvf: Update auto-generated files

Signed-off-by: Mario Maz <mmazzapater@gmail.com>

mariomaz committed:
agent: replace multiap_attr with bss_type byte

Modify reading and writing of cConfigData used in vendor specific
message ACTION_APMANAGER_WIFI_CREDENTIALS_UPDATE_REQUEST so it does not
use the struct sWscAttrVendorExtMultiAp anymore but the new bss_type
byte (this struct is wrong and is going to be removed in a future
commit).

Signed-off-by: Mario Maz <mmazzapater@gmail.com>

mariomaz committed:
tlvf: Update auto-generated files

Signed-off-by: Mario Maz <mmazzapater@gmail.com>

mariomaz committed:
tlfv: use new WSC Vendor Extension class

We currently have multiple attributes and classes all with the same
attribute type - vendor extension attribute - `sWscAttrVersion2`,
`sWscAttrVendorExtMultiAp`, `cWscVendorExtWfa`, and
`cWscAttrVersion2`. All are basically wrong since they do not support
the generic vendor extension attribute which does not necessarily have
to be WFA vendor extension.

Modify config_data, m1 and m2 messages to use the new WSC Vendor
Extension class.

Add Version2 subelement as the first subelement in the config data
WFA Vendor Extension where the second subelement is the MultiAP
Identifier subelement according to the WSC specification.

Signed-off-by: Mario Maz <mmazzapater@gmail.com>

mariomaz committed:
tlvf: Update auto-generated files

Signed-off-by: Mario Maz <mmazzapater@gmail.com>

rmelotte commented:
I added a commit to run MAP-4.10.4 again with the Qualcomm controller.
Since it's also failing on master (with the Mediatek controller) I don't expect it to pass, but we can at least manually check that we now correctly parse 1 fronthaul and 1 backhaul.

rmelotte commented:
The certification test currently fails here: 
```
DEBUG 21:44:17:666 <1998507040> son_slave_thread.cpp[4039] --> Received AP_AUTOCONFIGURATION_WSC_MESSAGE
ERROR 21:44:17:667 <1998507040> m2.cpp[313] --> TLVF: getAttr<cWscAttrVendorExtension> failed
ERROR 21:44:17:668 <1998507040> m2.cpp[32] --> TLVF: Not all attributes present
INFO 21:44:17:668 <1998507040> son_slave_thread.cpp[4045] --> Not a valid M2 - Ignoring WSC CMDU
```

rmelotte commented on framework/tlvf/src/include/tlvf/WSC/configData.h:
```diff
@@ -60,7 +60,36 @@ class configData : public AttrList {
+            }
+
+            // Skip this subelement
+            index += vendor_data[index];
```
```suggestion
            index += vendor_data[index] + 1;
```

arnout commented on framework/tlvf/src/include/tlvf/WSC/configData.h:
```diff
@@ -60,7 +60,36 @@ class configData : public AttrList {
-    uint8_t bss_type() const { return getAttr<cWscVendorExtWfa>()->subelement_value(); };
+    eWscVendorExtSubelementBssType bss_type() const
+    {
+        eWscVendorExtSubelementBssType default_value = eWscVendorExtSubelementBssType::TEARDOWN;
```
```suggestion
        constexpr eWscVendorExtSubelementBssType default_value = eWscVendorExtSubelementBssType::TEARDOWN;
```

mariomaz replied:
Done

arnout commented on framework/tlvf/src/include/tlvf/WSC/configData.h:
```diff
@@ -60,7 +60,36 @@ class configData : public AttrList {
+        if ((eWscVendorId::WSC_VENDOR_ID_WFA_1 != vendor_ext_attr->vendor_id_0()) ||
+            (eWscVendorId::WSC_VENDOR_ID_WFA_2 != vendor_ext_attr->vendor_id_1()) ||
+            (eWscVendorId::WSC_VENDOR_ID_WFA_3 != vendor_ext_attr->vendor_id_2())) {
+            return default_value;
```
To be fully correct, we should actually iterate over all vendor extension attributes until we find one that matches the WFA ID.

mariomaz replied:
Done

arnout commented on framework/tlvf/src/include/tlvf/WSC/configData.h:
```diff
@@ -60,7 +60,36 @@ class configData : public AttrList {
+        auto vendor_data = vendor_ext_attr->vendor_data();
+        while ((index + sizeof(sWscWfaVendorExtSubelementMultiApIdentifier)) <=
+               vendor_ext_attr->vendor_data_length()) {
+            if (eWscWfaVendorExtSubelement::MULTI_AP_IDENTIFIER == vendor_data[index++]) {
```
(nitpick) I heavily dislike the use of ++ like this. I prefer explicit indexing, i.e. keep index constant within one iteration and update it at the end of the loop body only.

Also, I prefer explicit variables (usually I don't like adding variables like that, but here I think it documents the meaning of the individual bytes):
```
auto subelement_type = vendor_data[index];
auto subelement_length = vendor_data[index + 1];
if (subelement_type == ... && subelement_length == 1) {
    return static_cast<WSC::eWscVendorExtSubelementBssType>(vendor_data[index + 2]);
}
```

mariomaz replied:
Done

arnout commented on framework/tlvf/src/include/tlvf/WSC/configData.h:
```diff
@@ -60,7 +60,36 @@ class configData : public AttrList {
+            }
+
+            // Skip this subelement
+            index += vendor_data[index++];
```
So this is why I don't like `index++`... It *looks* like index is incremented 3 times at first sight, you have to really follow the flow to understand that this is `vendor_data[index + 1]`...
```suggestion
            index += 2 + subelement_length;
```

mariomaz replied:
Done

arnout commented on framework/tlvf/src/src/WSC/m1.cpp:
```diff
@@ -204,21 +204,24 @@ bool m1::init(const config &cfg)
+    const size_t vendor_data_size = sizeof(sWscWfaVendorExtSubelementVersion2);
+    if (!vendor_ext_attr->alloc_vendor_data(vendor_data_size)) {
+        LOG(ERROR) << "Failed to allocate vendor data [" << vendor_data_size << "]!";
         return false;
     }
+    auto vendor_data = vendor_ext_attr->vendor_data();
 
-    sWscAttrVersion2 version2;
-    version2.struct_swap();
-    uint8_t *version2_buf = reinterpret_cast<uint8_t *>(&version2);
-    std::copy(version2_buf, version2_buf + sizeof(version2), vendor_ext_attr->vs_data());
+    // WFA Vendor Extension Subelement at #0: Version2
+    sWscWfaVendorExtSubelementVersion2 version2{eWscWfaVendorExtSubelement::VERSION2, 0x01,
+                                                eWscVendorExtVersionIE::WSC_VERSION2};
+    std::copy_n(reinterpret_cast<uint8_t *>(&version2), sizeof(version2), vendor_data);
```
Since this is used both in M1 and M2, it could be worth factoring it in a function (nitpick).

arnout reviewed (approved):
Feel free to ignore all of my comments (though I really would like to fix that index++ stuff, I had to read it four times before I was certain the code was indeed correct...)

mariomaz committed:
tlvf: add getAttrList() method

Currently there is a getAttr<T>() method in class AttrList to get an
attribute of a given type.

However, message can include several Vendor Extension attributes so we
need a AttrList::getAttrList<T>() method to get all of them.

Later we will iterate over all Vendor Extension attributes in the list
until we find one that matches the WFA ID.

Signed-off-by: Mario Maz <mmazzapater@gmail.com>
